NDTensors/src/UnallocatedArrays/src/unallocatedzeros.jl

```diff
@@ -0,0 +1,143 @@
+struct UnallocatedZeros{ElT,N,Axes,Alloc<:AbstractArray{ElT,N}} <:
+       FillArrays.AbstractZeros{ElT,N,Axes}
+  z::FillArrays.Zeros{ElT,N,Axes}
+  function UnallocatedZeros{ElT,N,Alloc}(inds::Tuple) where {ElT,N,Alloc}
+    z = FillArrays.Zeros{ElT,N}(inds)
+    Axes = typeof(FillArrays.axes(z))
+    return new{ElT,N,Axes,Alloc}(z)
+  end
+
+  function UnallocatedZeros{ElT,N,Alloc}(::Tuple{}) where {ElT,N,Alloc}
+    @assert N == 1
+    z = FillArrays.Zeros{ElT,N}(1)
+    Axes = typeof(FillArrays.axes(z))
+    return new{ElT,N,Axes,Alloc}(z)
+  end
+
+  function UnallocatedZeros{ElT,N,Axes,Alloc}(inds::Tuple) where {ElT,N,Axes,Alloc}
+    @assert Axes == typeof(Base.axes(inds))
+    z = FillArrays.Zeros{ElT,N}(inds)
+    return new{ElT,N,Axes,Alloc}(z)
+  end
+end
+
+function UnallocatedZeros{ElT,N,Axes,Alloc}() where {ElT,N,Axes,Alloc<:AbstractArray{ElT,N}}
+  return UnallocatedZeros{ElT,N,Axes,Alloc}[]
+end
+function UnallocatedZeros(alloc::Type{<:AbstractArray}, inds...)
+  @assert ndims(alloc) == length(inds)
+  alloc = specify_eltype(alloc)
+  return UnallocatedZeros{eltype(alloc),ndims(alloc),alloc}(Tuple(inds))
+end
+
+function UnallocatedZeros{ElT}(alloc::Type{<:AbstractArray}, inds...) where {ElT}
+  alloc = set_eltype(alloc, ElT)
+  N = length(Tuple(inds))
+  alloc = set_ndims(alloc, N)
+  return UnallocatedZeros(alloc, inds...)
+end
+
+Base.ndims(::UnallocatedZeros{ElT,N}) where {ElT,N} = N
+ndims(::UnallocatedZeros{ElT,N}) where {ElT,N} = N
+Base.eltype(::UnallocatedZeros{ElT}) where {ElT} = ElT
+alloctype(::UnallocatedZeros{ElT,N,Axes,Alloc}) where {ElT,N,Axes,Alloc} = Alloc
+function alloctype(::Type{<:UnallocatedZeros{ElT,N,Axes,Alloc}}) where {ElT,N,Axes,Alloc}
+  return Alloc
+end
+axes(::Type{<:UnallocatedZeros{ElT,N,Axes}}) where {ElT,N,Axes} = Axes
+
+Base.size(zero::UnallocatedZeros) = Base.size(zero.z)
+
+Base.print_array(io::IO, X::UnallocatedZeros) = Base.print_array(io, X.z)
+
+data(zero::UnallocatedZeros) = zero.z
+getindex(zero::UnallocatedZeros) = getindex(zero.z)
+
+array(zero::UnallocatedZeros) = alloctype(zero)(zero.z)
+Array(zero::UnallocatedZeros) = array(zero)
+axes(z::UnallocatedZeros) = axes(z.z)
+dims(z::UnallocatedZeros) = Tuple(size(z.z))
+dim(z::UnallocatedZeros) = dim(size(z.z))
+copy(z::UnallocatedZeros) = UnallocatedZeros{eltype(z),ndims(z),alloctype(z)}(dims(z))
+
+Base.vec(z::Type{<:UnallocatedZeros}) = z
+Base.vec(z::UnallocatedZeros) = z
+
+function Base.convert(x::Type{T}, z::UnallocatedZeros) where {T<:Base.Array}
+  return Base.convert(x, z.z)
+end
+
+function complex(z::UnallocatedZeros)
+  ElT = complex(eltype(z))
+  N = ndims(z)
+  AllocT = similartype(alloctype(z), ElT)
+  return UnallocatedZeros{ElT,N,AllocT}(dims(z))
+end
```

Comment:
Seems like there should be a more elegant code pattern for functions like this, for example:
```julia
function complex(z::UnallocatedZeros)
  return set_eltype(typeof(z), complex(eltype(z))
)(complex(parent(z)))
end
```

Reply:
Or:
```julia
set_alloctype(complex(parent(Z)), set_eltype(alloctype(Z), complex(eltype(Z)))
```

Reply:
I have this implemented now and it works. Thanks!